### ISSUE TITLE
Golang network filter missing functions

### ISSUE THREAD

**Issue by or-adar** [envoyproxy/envoy#29270]
I'm experimenting with the new golang network filter in v1.27.0
and I noticed that native C++ filters are exposed to more functions than what the golang filter is exposing.
for instance, the ConnectionCallback interface: https://github.com/envoyproxy/envoy/blob/f17f1fdd644cc4b32f65579184757e3975843db6/contrib/golang/common/go/api/filter.go#L176-L183
exposes 3 methods,
but from what it looks like the connection callback should have more methods such as `requestedServerName()` to retreive the SNI of the requested server and so on..

so, am I missing something? or is there a plan to add any missing methods to the connection callback later on?

**Comment by doujiang24** [comment]
Yep, the network filter is the early iteration, new features in the TODO list.

Could you please describe your case a bit, and is there other APIs you needed? except the `requestedServerName`.

cc @antJack

**Comment by doujiang24** [comment]
> is there a page to track the TODO list you mentioned?

not yet, feel free to create issues, or even PRs, when you need new features.

> And for my use case, I wanted to use requestedServerName to examine some custom behaviour that will mint a self signed certificate in the bacground based on the SNI on the request,

Do you want to generate a self signed certificate, on demand, by using Golang, based on the SNI?
if yes, I think it might be too late?

**Comment by or-adar** [comment]
I was planning to use the generated self signed certificate on subsequent requests with that SNI using xDS,
so it could be late but not that late.

**Comment by doujiang24** [comment]
@or-adar oh, I see. then, how does first SSL/TLS connection handshake? is there already a useable bootstrap certificate?

**Comment by or-adar** [comment]
@doujiang24 No, the first connection is just forwarded as-is and the TLS handshake won't be handled by Envoy in that case.

but coming to think of it, maybe it won't be really great for what I'll be trying to acheive.
do you think I should use a different approach? like creating a custom handshaker that'll generate certificates for each domain that a user has requested at runtime? (if custom_hanshaker is the case, do you know of any good examples that use it?)

**Comment by doujiang24** [comment]
yep, custom_hanshaker could be a workable choice, but it's a bit overhead if we just need generate certificates at runtime.
it will take over the SSL session resumption and SSL ticket, and etc.

actually, I had implemented a custom_handshaker that support using Golang to hand generate certificate at runtime, but, not opensource yet. I might create a PR for it when I have enough free time.

But, I do think a better choice, might be a `custom_certificate_specifier`, similar to `custom_handshaker`, so that the golang extension that implemented the `custom_certificate_specifier` do not need to care about SSL session/ticket.

**Comment by or-adar** [comment]
it could produce a bit overhead but it can be tolerated, as theoretically I will only need to generate certificates only once (per domain), and reuse existing certificates for consequent requests with that same domain.

I haven't heard about `custom_certificate_specifier` or got to find anything about it in the docs/codebase,
do you have any reference?

Also, does creating `custom_certificate_specifier`/`custom_handshaker` requires forking the project? 
or you can write the custom logic in a separate codebase and reference it with a compiled binary that you mount to a container running envoy (similar to how it could be acheived with wasm/golang network filter)?
I just never seen any examples of `custom_handshaker` so if you have any known exaples or references I would be happy if you could share them!

**Comment by doujiang24** [comment]
> just to be clear, the golang extension for `custom_certificate_specifier/custom_handshaker` is something that is planned?

yep, it's on our TODO list, but not a high priority. anyway, I'll create a PR for the draft version.

> And the PR you are talking about showcases a golang extension that allows extending `custom_certificate_specifier/custom_handshaker`?

yep, that's it, but it might not be a fully workable stage, might be a draft version at first.

**Comment by or-adar** [comment]
oh I see, thank you for clarifying!
Looking forward to seeing the progress

**Comment by BenAgai** [comment]
Hi @or-adar, @doujiang24 
I’m currently researching how to do selective TLS bumping that will affect incoming connections.
i.e hold incoming connection, accepted by Envoy, before TLS handshake is made, generate certificate for them if needed and perform TLS bumping.

I, too, didn’t find a way to do the above. I succeeded in implementing selective TLS bumping but it won’t work with newly incoming connections for which I didn’t create certificate yet (as the certificate creation happens at a later point where the handshake was already done).

@doujiang24 ,
Regarding your response on using custom_handshaker, I would like to avoid code changes in Envoy as much as possible.
Sorry if I’m asking the same question, but is there any plan to provide Go API to do the above without the need to change Envoy’s code base? (A POC/draft will be good to look at as well)

Thanks!

**Comment by doujiang24** [comment]
> is there any plan to provide Go API to do the above without the need to change Envoy’s code base?

yep, just like the current Golang filter extension: https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/golang_filter

**Comment by doujiang24** [comment]
Hello @or-adar and @BenAgai , here is the draft PR: https://github.com/envoyproxy/envoy/pull/29383

But I may not have enough time to finish it in short, welcome to continue it, or other better ways, if you are instersted it.

**Comment by BenAgai** [comment]
Thanks @doujiang24  for the reply!
I hope #29383 will be prioritized as I believe the above feature is desired by more people.

Have a good week!
